### ISSUE TITLE
[Document view] : Need to correctly handle ObjectID-like data when specified in nested fields

### ISSUE THREAD

**Issue by PrashanthCorp** [Microsoft/vscode-cosmosdb#564]
Courtesy @Danlock for pointing this out. Pasting comment below: 

Prior discussion [here](https://github.com/Microsoft/vscode-cosmosdb/issues/534#issuecomment-383963490)

Related to this issue, nested objectId are not set as object ids properly, in the case of:

```
{
  "_id": "5aa0150e1430142dda49b551",
  "createdAt": "2018-03-07T16:36:30.436Z",
  "data": {
    "Quasi": "8590406b-67fa-4e08-819d-2860f5fdca79"
  },
  "name": "g6d-car",
  "sensors": [
    {
      "_id": "5adf3b43194dd500010542ea",
      "createdAt": "2018-04-24T14:12:19.693Z",
      "type": "camera"
    }
  ],
  "type": "car",
  "updatedAt": "2018-04-24T14:12:19.693Z"
}
```

If I make any changes the inner sensor ids will be overwritten with "5adf3b43194dd500010542ea" instead of ObjectId("5adf3b43194dd500010542ea").

I can make another issue if you want, but this seems relevant.

**Comment by Danlock** [comment]
@PrashanthCorp  Thanks for looking out, but that much info is fine. Mostly autogenerated.

I also want to stress that the key, "_id" doesn't really matter, the objectid value ObjectId("5adf3b43194dd500010542ea") is being overwritten and thats the problem, even if the keys isn't "_id".
For example a nested field like so

```
{
  "_id": "5aa0150e1430142dda49b551",
  "createdAt": "2018-03-07T16:36:30.436Z",
  "data": {
    "Quasi": "8590406b-67fa-4e08-819d-2860f5fdca79"
  },
  "name": "g6d-car",
  "sensors": [
    {
      "_id": "5adf3b43194dd500010542ea",
      "createdAt": "2018-04-24T14:12:19.693Z",
     "foreignIDs": ["5adf3b43194dd500010542ea","5adf3b43194dd500010542ea"]
      "type": "camera"
    }
  ],
  "type": "car",
  "updatedAt": "2018-04-24T14:12:19.693Z"
}
```

They would also need to be persisted as ObjectId("5adf3b43194dd500010542ea") instead of "5adf3b43194dd500010542ea" to avoid corruption.

Thanks a lot for your work, for reading data this plugin is very convenient to use.

**Comment by Danlock** [comment]
I see the issue. That means this issue probably also applies to ISODate() and other mongoDB types too.

If you guys are married to strict JSON output then you could peek into every value for certain prefixes then stringify the whole thing 
` ISODate("2018-04-24T14:12:19.693Z") -> "ISODate("2018-04-24T14:12:19.693Z")" `

**Comment by dancherb** [comment]
I'm just come here after spending the day figuring out why my back-end was suddenly broken and all sorts of things in my app (commenting, up-voting etc.) were no longer working. It's because I tried out the extension yesterday and opened the document with cosmos DB - all of the object ID's in my docs were turned to literal strings, and my MongoDB/mongoose operations were no longer finding the docs.

Lucky I was still in development, but if I wasn't, using Cosmos DB would have completely broken my MongoDB database and app. Serious stuff! Is this a recent issue and I just got unlucky? Seems like it was only posted today (and @danlock even commented while I was typing this out).

**Comment by StephenWeatherford** [comment]
The result JSON has always been this way as far as we know, even in the original mongo extension that later became this extension.  However, I think the update functionality is newer, but still at least several months old without any comments.

Sorry for the inconvenience, and for letting us know about the issue and the seriousness of it. We were not aware of the possibility of data loss and will definitely be looking into how to fix this.

**Comment by PrashanthCorp** [comment]
I have a fix for the issue. See above PR. 

For test, I tried the following document: 

```
{
  "_id": {
    "$oid": "5abeb30758236b43548ba89a"
  },
  "relatedTo": [
    {
      "$oid": "5a8ca0b0918a2f3c14611736"
    },
    {
      "$oid": "5abeb30758236b43548ba89c"
    }
  ],
  "createdAt": {
    "$date": "2018-04-24T14:12:19.693Z"
  },
  "sampleTime": {
    "$timestamp": {
      "t": 0,
      "i": 0
    }
  },
  "parent": {
    "$undefined": true
  },
  "sampleMinKey": {
    "$minKey": 1
  },
  "sampleMaxKey": {
    "$maxKey": 1
  },
  "entree": {
    "name ": "sushi",
    "base": "rice",
    "grainsOfRiceUsed": {
      "$numberLong": "9223372036854775807"
    },
    "weightInGrams": {
      "$numberDecimal": "80.321"
    }
  }
}
```

Let me know if there's any major cases I'm missing. 

Let us know if you'd like a pre-rel vsix. That could help you work on your databases.

**Comment by Danlock** [comment]
there is also a UUID() constructor, and undefined is deprecated so it's probably unnecessary to test for.

Also should probably test for arrays of these values. Other than that looks good enough for me.

**Comment by PrashanthCorp** [comment]
I have a basic test for arrays under the `relatedTo` property. 

From my understanding, UUID is part of the binary data type, of sub-type 3 or 4 [here](https://docs.mongodb.com/v3.2/reference/method/UUID/#example) and [in the spec - under non-terminals](http://bsonspec.org/spec.html). 

For testing that the constructed UUID remains consistent, I used the mongo shell to create a UUID(with a 32 digit hex seed), insert that in a document, and verify that the uuid property (24 chars, different from the 32 digit seed) is consistent across out extension and the shell. Run a few updates on the doc in the extension, and the uuid remains unchanged. 
For basis, I used the seed from the [mongo docs] (https://docs.mongodb.com/v3.2/reference/method/UUID/). The generated UUID matches the example through the updates. 

Hope this helps!
